src/encrypt/des_ecb.c

```diff
@@ -0,0 +1,185 @@
+ * software and associated documentation files (the "Software"), to deal in the Software
+ * without restriction, including without limitation the rights to use, copy, modify,
+ * merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <common.h>
+#include <stdlib.h>
+#include <string.h>
+
+/**
+ * Generate an DES key with a template suitable for encrypting data.
+ * The key is a Session key, and will be deleted once the HSM Session is closed.
+ * @param session Active PKCS#11 session
+ * @param key Location where the key's handle will be written
+ * @return CK_RV
+ */
+CK_RV generate_des_key(CK_SESSION_HANDLE session,
+                       CK_OBJECT_HANDLE_PTR key) {
+    CK_MECHANISM mech;
+
+    mech.mechanism = CKM_DES3_KEY_GEN;
+    mech.ulParameterLen = 0;
+    mech.pParameter = NULL;
+
+    CK_ATTRIBUTE template[] = {
+            {CKA_TOKEN,       &false_val,            sizeof(CK_BBOOL)},
+            {CKA_EXTRACTABLE, &true_val,             sizeof(CK_BBOOL)},
+            {CKA_ENCRYPT,     &true_val,             sizeof(CK_BBOOL)},
+            {CKA_DECRYPT,     &true_val,             sizeof(CK_BBOOL)},
+    };
+
+    return funcs->C_GenerateKey(session, &mech, template, sizeof(template) / sizeof(CK_ATTRIBUTE), key);
+}
+
+/**
+ * Encrypt and decrypt a string using DES ECB.
+ * @param session Active PKCS#11 session
+ */
+CK_RV des_ecb_sample(CK_SESSION_HANDLE session) {
+    CK_RV rv;
+
+    // Generate a DES key.
+    CK_OBJECT_HANDLE des_key;
+    rv = generate_des_key(session, &des_key);
+    if (CKR_OK != rv) {
+        printf("DES key generation failed: %lu\n", rv);
+        return rv;
+    }
+
+    CK_BYTE_PTR plaintext = "Data must be a 16 byte multiple.";
+    CK_ULONG plaintext_length = strlen(plaintext);
+    CK_ULONG ciphertext_length = 0;
+
+    printf("Plaintext: %s\n", plaintext);
+    printf("Plaintext length: %lu\n", plaintext_length);
+
+    // Prepare the mechanism
+    CK_MECHANISM mech = {CKM_DES3_ECB, NULL, 0};
+
+    //**********************************************************************************************
+    // Encrypt
+    //**********************************************************************************************
+
+    rv = funcs->C_EncryptInit(session, &mech, des_key);
+    if (CKR_OK != rv) {
+        printf("Encryption Init failed: %lu\n", rv);
```

Comment:
Could you update the failure cases to be `fprintf(stderr, ...)`?

Reply:
resolved.